pineappl/src/grid.rs

```diff
@@ -1595,6 +1613,20 @@ impl Grid {
                             0
                         };
 
+                        // if `op1` and `op2` below are zero there's no work to do
+                        // TODO: ideally we change the for loops instead of vetoing here
+                        if (has_pdf1
+                            && !non_zero_pid_indices
+                                .iter()
+                                .any(|&tuple| tuple == (tgt_pid1_idx, src_pid1_idx)))
+                            || (has_pdf2
+                                && !non_zero_pid_indices
+                                    .iter()
+                                    .any(|&tuple| tuple == (tgt_pid2_idx, src_pid2_idx)))
+                        {
+                            continue;
```

Comment:
@cschwan if we assume for a second the error is in here, this `continue` is realistically the only option for something going wrong. So,
- the comment says "if `op1` **and** `op2`", but the condition features an **or** between two things - does this make sense?
- the other option is `non_zero_pid_indices` being computed wrongly or used wrongly, which could explain why some grids are working and others not

I don't believe this to be a problem, but "the x1 and x2 grid values are the same across all non-zero subgrids" - is this statement ever asserted?

cc @andreab1997

Reply:
> * the comment says "if `op1` **and** `op2`", but the condition features an **or** between two things - does this make sense?

Actually, I guess the comment is wrong: it should be "if either `op1` or `op2` is zero there's no work to do"

Reply:
> the other option is non_zero_pid_indices being computed wrongly

now that I'm thinking about it: @andreab1997 another possible source of a bug might be the flavor rotation inside pineko: because we rotate there the eko from flavor-flavor to evolution-flavor as FK tables are by default in evolution space, but eko and grids are by default in flavor space only (actually, eko is a tensor so it has to set of indices). But again since it is working for some datasets this is unlikely. However, this would a possibility of eko being correct, yet FK wrong.

Reply:
> now that I'm thinking about it: @andreab1997 another possible source of a bug might be the flavor rotation inside pineko: because we rotate there the eko from flavor-flavor to evolution-flavor as FK tables are by default in evolution space, but eko and grids are by default in flavor space only (actually, eko is a tensor so it has to set of indices). But again since it is working for some datasets this is unlikely. However, this would a possibility of eko being correct, yet FK wrong.

Yes indeed is unlikely but I can try to test it.

Reply:
@felixhekhorn the next commit adds negations, and I think it should be an OR instead of an AND, at least for the double hadronic case. Can you confirm or deny this!

Reply:
I think that condition might be fine as is - as said above it's only the comment which is wrong - moreover, we just tried to just remove the `continue` and it seems nothing changes, i.e. this might not be the offending change - unfortunately ... (actually, even in practice the condition should be almost always `false`). However, now we also suspect that even DIS might be affected (@andreab1997 can maybe confirm) which is easier to check.

Reply:
Actually I just tested NMCPD_dw_ite (only NMC_NC_EM_P_F2.pineappl.lz4) and it worked. I am going to try also the other DIS fks but at the moment I would say that the bug does not affect DIS.

Reply:
I have tried about 10 different DIS FKs and they are all correct with pineko==0.2.2 and pineappl==0.5.5

Reply:
I'm now convinced that commit b7014fdc4ecf9ed3ef15f08cc47ee5761ed0b471 introduces a bug in the case where the x-grids of processes change over bins, orders or luminosities. This usually happens in the case we convert them from old APPLgrids. @andreab1997 do you a need new release to test a different version of PineAPPL or can you also test it from git?

Reply:
Since it has been modified a bit, I'd propose to revert it manually in a new branch, and test if that's the issue.

Reply:
I can test it from git but today I am a bit busy with seminars. If you can test it thank you, otherwise I will test it soon as possible.

Reply:
I will do the commit, and @felixhekhorn has a quick test already in place

Reply:
Alright, thanks to everyone for figuring this out! I will mark this discussion as resolved, and everything else we'll discuss in #182.